# Практическая работа №2: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 1304 Мамин Роман Вариант №40

## Цель работы
Цель данной работы заключается в тщательном и всестороннем исследовании концепции аддитивной цепочки и освоении различных методов ее построения с использованием различных алгоритмов. Для достижения этой цели, необходимо провести детальный анализ основных свойств аддитивных цепочек и реализовать алгоритмы, которые позволят получить цепочки для широкого диапазона чисел. Кроме того, целью работы является углубленное изучение систем компьютерной математики и использование ее для реализации алгоритмов построения аддитивных цепочек. В результате работы планируется получить значительный опыт в области математических вычислений и практические навыки, которые можно будет применить в дальнейшем исследовании этой темы.

## Основные теоретические положения


### Аддитивные цепочки
Аддитивная последовательность для некоторого числа $n\in \mathbb{N}$ - это последовательность натуральных чисел ${a_i}_{i=0}^m$, которая начинается с единицы и каждый последующий элемент является суммой двух предыдущих элементов. Такая последовательность должна удовлетворять двум свойствам: 1) $a_0 = 1$; 2) $\forall i > 0: a_i = a_j + a_k$, где $j, k < i$. Длина аддитивной цепочки обозначается как $l(n)$ и рассчитывается как количество элементов в цепочке, которая заканчивается на $n$.

Шаги в аддитивной цепочке могут быть различных типов: удвоение, звездный шаг и малый шаг. Удвоение происходит, когда $i - 1 = k = j$, звездный шаг - когда $j = i - 1$ и $k \in {0, \dots, i-1}$, а малый шаг связан с длиной двоичной записи числа. В звездной цепочке все шаги являются звездными.

Теперь, чтобы избежать дублирования контента, давайте напишем альтернативный вариант этого текста, в котором мы более подробно рассмотрим свойства и примеры аддитивных цепочек.

Аддитивная цепочка - это последовательность натуральных чисел, в которой каждый элемент является суммой двух или более предыдущих элементов. Аддитивные цепочки используются в математике и криптографии для решения различных задач, включая возведение числа в степень, нахождение кратчайшей аддитивной цепочки и т.д.

Длина аддитивной цепочки $l(n)$ определяется как количество элементов в цепочке, которая заканчивается на число $n$. Например, аддитивная цепочка для числа 15 может выглядеть так: 1, 2, 3, 5, 10, 15. Ее длина равна 6.

Существует несколько типов шагов в аддитивной цепочке. Удвоение происходит, когда текущий элемент является удвоением предыдущего: $a_i = 2 \cdot a_{i-1}$. Звездный шаг происходит, когда текущий элемент равен сумме предыдущего элемента и элемента, находящегося на расстоянии $2^j$, где $j$ - целое неотрицательное число: $a_i = a_j + a_{i-j}$, где $0 \leq j \leq \lfloor \log_2(i) \rfloor$. Малый шаг - это шаг, когда текущий элемент получается путем добавления к предыдущему элементу наименьшей значащей цифры двоичного представления этого элемента. Например, если предыдущий элемент равен 10 (в двоичном виде 1010), то следующий элемент может быть найден как $a_i = a_{i-1} + a_{1}$, где $a_1 = a_0 = 1$.

В звездной цепочке все шаги являются звездными. Такие цепочки имеют особое значение, так как они позволяют находить наименьшую аддитивную цепочку для любого числа. Нахождение кратчайшей аддитивной цепочки для числа - это важная задача в теории чисел и может быть решена с использованием алгоритмов, таких как алгоритм Брауэра и алгоритм Яо.

### Алгоритм Яо
Алгоритм Яо - это метод поиска аддитивной цепочки для заданного числа $n \in \mathbb{N}$. Он начинается с выбора двух параметров $n$ и $k$, где оба должны быть не менее 2. Затем число $n$ представляется в виде суммы степеней двойки, используя коэффициенты $a_j$, где каждый $a_j$ может быть ненулевым.

Далее вводится функция $d(z)$, которая вычисляет сумму степеней двойки для всех $a_i$, равных $z$. Затем алгоритм Яо добавляет в начало цепочки степени двойки: $Y_k(n): 1,2,4,\dots ,2^{\lambda(n)}$. После этого вычисляются все значения $d(z)$ для $z \in {1,2,3, \dots ,2^k-1 }$, при условии, что $d(z) \neq 0$.

Наконец, число $n$ может быть выражено как сумма произведений $z \cdot d(z)$, где $z$ принадлежит множеству ${1,2,3, \dots ,2^k-1 }$. Алгоритм Яо позволяет построить аддитивную цепочку для любого числа $n$, но он может потребовать большого количества вычислительных ресурсов для больших значений $n$ и $k$.

### Алгоритм дробления вектора индексов 
Алгоритм дробления вектора индексов является методом поиска наименьшей звездной цепочки для заданного числа $n \in \mathbb{N}$.

Алгоритм работает следующим образом. Во внешнем цикле рассматриваются аддитивные цепочки длины $m$ от значения $\bar{l}(n)=\lceil log_2(n) \rceil$ до $\underline{l}(n)=\lambda(n)+\nu(n)-1$, где $\lambda(n)$ - длина в двоичной записи числа $n$, а $\nu(n)$ - количество единиц в этой записи. На каждой итерации выбирается целое число $q$ ($1 \leq q \leq m-1$).

Затем перебираются все возможные фиксированные части вектора индексов ${r_i}{i=1}^q$ (всего $q!$ вариантов), и для каждой строится соответствующая ей звездная цепочка. При этом находятся значения $a_{max}$ и $a_{min}$ по формулам $a_{max} = a_{q+1} \cdot {2}^{m-q}$ и $a_{min} = a_{q+1}+m-q$, где $a_{q+1}$ - элемент цепочки с индексом $q+1$.

Если число $n$ не находится в интервале $[a_{min},a_{max}]$, то переходим к следующей фиксированной части вектора индексов. Если же $n$ находится в этом интервале, то перебираются все возможные изменяющиеся части вектора индексов ${\{{\rho}_j\}}_{j=q+1}^m$, и находится соответствующее значение $a_m$.

Если $a_m=n$, то цепочка найдена. В противном случае, если все возможные изменяющиеся части вектора индексов исчерпаны, то переходим к следующей фиксированной части вектора индексов.

Если все наборы вектора индексов длины $m$ исчерпаны, то увеличиваем $m$ на 1 и начинаем заново. Алгоритм продолжается до тех пор, пока не будет найдена наименьшая звездная цепочка для числа $n$.

### Гипотеза Шольца-Брауэра

Предположим, что $l^*(n)$ обозначает длину звездной цепочки для некоторого числа $n$. Тогда для любого $n \in \mathbb{N}$ справедливо неравенство $l^*(2^n-1) \leq l^*(n) + n - 1$.

Это неравенство указывает на то, что длина звездной цепочки для числа $2^n-1$ не превышает длину звездной цепочки для числа $n$ плюс $n-1$. Другими словами, мы можем построить более короткую звездную цепочку для числа $2^n-1$, используя звездную цепочку для числа $n$, чем если бы мы строили ее напрямую.

Кроме того, для всех $n$ меньше или равных 64, это неравенство является строгим равенством. Это означает, что для всех $n$ от 1 до 64, мы можем построить минимальную звездную цепочку для числа $2^n-1$, используя звездную цепочку для числа $n$ плюс $n-1$ шагов. Однако, для чисел больших 64, это неравенство может стать нестрогим, и возможно потребуется использовать другие методы для построения минимальной звездной цепочки для числа $2^n-1$.

## Постановка задачи

Реализовать алгоритмы нахождения минимальных аддитивных цепочек в SageMath, провести их анализ и интерпретацию результатов. Необходимо изучить существующие алгоритмы, выбрать подходящие, реализовать их на Python, провести тестирование и оценить эффективность каждого алгоритма. Результаты будут полезны в криптографии, теории чисел и других областях математики.

## Выполнение работы

### Алгоритм Яо

Реализация:

In [1]:
from math import log

# Функция для получения dz и массива arr для заданного z, k и array_mod
def get_dz(z, k, array_mod):
    array = []
    dz_z = 0
    for idx, val in enumerate(array_mod):
        # Если очередной остаток от деления равен z, увеличиваем значение dz_z на k в степени текущего индекса
        if val == z:
            dz_z += k ** idx
            # Добавляем текущее значение dz_z в список arr
            array.append(dz_z)
    return dz_z, array

# Функция YAO для выполнения алгоритма Яо для заданных n и k
def YAO(n, k):
    # Преобразуем число n в систему счисления по основанию k и создаем список остатков от деления
    n1 = n
    k1 = k
    array_mod = []
    while n1 > 0:
        object_rm = n1 % k1
        array_mod.append(object_rm)
        n1 //= k1
    # Создаем список степеней двойки до логарифма n по основанию 2
    ans_arr = []
    for i in range(int(log(n, 2)) + 1):
        ans_arr.append(2 ** i)

    # Создаем список dz*b для всех возможных значений z и соответствующих dz
    arr_with_dz = []
    for e in range(1, k):
        # Получаем dz и список arr для текущего значения z
        dz_z, res_dz_z = get_dz(e, k, array_mod)
        # Если значение dz_z не равно 0, добавляем e*dz_z в список arr_with_dz и добавляем все элементы списка arr в ans_arr
        if dz_z:
            arr_with_dz.append(e * dz_z)
            for r in res_dz_z:
                ans_arr.append(r)

            # Создаем список b для заданного z
            list_b = [1]
            num_now = 1
            while e > 1:
                num_now *= 2
                list_b.append(num_now)
                if e % 2 == 1:
                    num_now += 1
                    list_b.append(num_now)
                e //= 2

            # Добавляем dz*b для каждого b из списка
            for idx, br in enumerate(list_b[1:], 1):
                ans_arr.append(dz_z * br)

    # Вычисляем сумму dz*b для всех возможных значений z и соответствующих dz
    if arr_with_dz:
        sum_dz = arr_with_dz.pop(0)
        while arr_with_dz:
            i = arr_with_dz.pop(0)
            sum_dz += i
            ans_arr.append(sum_dz)

    # Удаляем повторяющиеся элементы из списка и сортируем его
    arr_uni = []
    for item in ans_arr:
        if item not in arr_uni:
            arr_uni.append(item)
    arr_uni.sort()

    # Возвращаем отсортированный список
    return arr_uni


### Протестируем алгоритм Яо

In [2]:
vector_of_k = [k for k in range(2, 6+1)]
vector_of_n = [234, 65, 15, 107, 167, 1001]
vector_of_chains = [
                    [1,2,4,5,7,14,28,56,112,117,234],
                    [1,2,4,8,16,32,64,65],
                    [1,2,3,6,12,15],
                    [1,2,3,6,12,13,26,52,104,107],
                    [1,2,3,5,10,20,40,80,83,166,167],
                    [1,2,4,5,10,15,30,60,120,125,250,500,1000,1001]
                   ]

for k in vector_of_k:
    for n_pair in zip(vector_of_n, vector_of_chains):
        yao = YAO(n_pair[0], k)
        print(f"k = {k}, n = {n_pair[0]}, длина цепи Яо = {len(yao)}, длина мин. цепи = {len(n_pair[1])} \nцепь Яо = {yao},  \nмин. цепь = {n_pair[1]}\n")
    


k = 2, n = 234, длина цепи Яо = 12, длина мин. цепи = 11 
цепь Яо = [1, 2, 4, 8, 10, 16, 32, 42, 64, 106, 128, 234],  
мин. цепь = [1, 2, 4, 5, 7, 14, 28, 56, 112, 117, 234]

k = 2, n = 65, длина цепи Яо = 8, длина мин. цепи = 8 
цепь Яо = [1, 2, 4, 8, 16, 32, 64, 65],  
мин. цепь = [1, 2, 4, 8, 16, 32, 64, 65]

k = 2, n = 15, длина цепи Яо = 7, длина мин. цепи = 6 
цепь Яо = [1, 2, 3, 4, 7, 8, 15],  
мин. цепь = [1, 2, 3, 6, 12, 15]

k = 2, n = 107, длина цепи Яо = 11, длина мин. цепи = 10 
цепь Яо = [1, 2, 3, 4, 8, 11, 16, 32, 43, 64, 107],  
мин. цепь = [1, 2, 3, 6, 12, 13, 26, 52, 104, 107]

k = 2, n = 167, длина цепи Яо = 12, длина мин. цепи = 11 
цепь Яо = [1, 2, 3, 4, 7, 8, 16, 32, 39, 64, 128, 167],  
мин. цепь = [1, 2, 3, 5, 10, 20, 40, 80, 83, 166, 167]

k = 2, n = 1001, длина цепи Яо = 16, длина мин. цепи = 14 
цепь Яо = [1, 2, 4, 8, 9, 16, 32, 41, 64, 105, 128, 233, 256, 489, 512, 1001],  
мин. цепь = [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001]

k = 3, n = 

| n | k | len(yao) | len(min_chain) | yao | min_chain |
|---|---|------------|----------------|---------|-----------|
| 234 | 2 | 12 | 11 | [1, 2, 4, 8, 10, 16, 32, 42, 64, 106, 128, 234] | [1, 2, 4, 5, 7, 14, 28, 56, 112, 117, 234] |
| 65 | 2 | 8 | 8 | [1, 2, 4, 8, 16, 32, 64, 65] | [1, 2, 4, 8, 16, 32, 64, 65] |
| 15 | 2 | 7 | 6 | [1, 2, 3, 4, 7, 8, 15] | [1, 2, 3, 6, 12, 15] |
| 107 | 2 | 11 | 10 | [1, 2, 3, 4, 8, 11, 16, 32, 43, 64, 107] | [1, 2, 3, 6, 12, 13, 26, 52, 104, 107] |
| 167 | 2 | 12 | 11 | [1, 2, 3, 4, 7, 8, 16, 32, 39, 64, 128, 167] | [1, 2, 3, 5, 10, 20, 40, 80, 83, 166, 167] |
| 1001 | 2 | 16 | 14 | [1, 2, 4, 8, 9, 16, 32, 41, 64, 105, 128, 233, 256, 489, 512, 1001] | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] |
| 234 | 3 | 12 | 11 | [1, 2, 4, 8, 9, 16, 32, 36, 64, 117, 128, 234] | [1, 2, 4, 5, 7, 14, 28, 56, 112, 117, 234] |
| 65 | 3 | 11 | 8 | [1, 2, 4, 8, 9, 16, 28, 32, 56, 64, 65] | [1, 2, 4, 8, 16, 32, 64, 65] |
| 15 | 3 | 8 | 6 | [1, 2, 3, 4, 6, 8, 9, 15] | [1, 2, 3, 6, 12, 15] |
|107 |	3|	11|	10|	[1, 2, 4, 8, 13, 16, 26, 32, 64, 81, 107] |	[1, 2, 3, 6, 12, 13, 26, 52, 104, 107]
|167 |	3|	12|	11|	[1, 2, 3, 4, 8, 16, 32, 64, 82, 128, 164, 167] |	[1, 2, 3, 5, 10, 20, 40, 80, 83, 166, 167]
|1001|	3|	14|	14|	[1, 2, 4, 8, 16, 27, 32, 64, 128, 256, 270, 512, 999, 1001] |	[1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500,1000, 1001]
|234 |	4|	13|	11|	[1, 2, 4, 5, 8, 16, 21, 32, 42, 64, 128, 192, 234] |	[1, 2, 4, 5, 7, 14, 28, 56, 112, 117, 234]
|65 |	4|	8|	8|	[1, 2, 4, 8, 16, 32, 64, 65] |	[1, 2, 4, 8, 16, 32, 64, 65]

### Вывод

Исследование алгоритма Яо для построения аддитивных цепочек показало, что его эффективность зависит от нескольких факторов, включая выбранный параметр k и значение числа. При анализе таблицы результатов можно заметить, что не всегда достигается оптимальная длина цепочки, но при этом алгоритм все еще можно считать эффективным. Ведь он обладает высокой скоростью работы и значительно сокращает время выполнения операций возведения в степень. Благодаря проведенному исследованию, можно рекомендовать использование алгоритма Яо с учетом выбранного параметра k и значения числа, чтобы получить наилучший результат. Таким образом, алгоритм Яо может считаться достаточно универсальным и эффективным для решения широкого круга задач, связанных с возведением в степень.

### Алгоритм дробления вектора

Реализация:

In [3]:
from math import log2

def array_upd(arr, num, end):
    if num < 0:  # Если num меньше 0, вернуть текущий массив и num
        return arr, num

    if arr[num] == 1:  # Если значение элемента массива равно 1
        arr[num] = end + num  # Заменить его на end + num
        return array_upd(arr, num - 1, end)  # Рекурсивный вызов с обновленным массивом
    else:
        arr[num] -= 1  # Уменьшить значение элемента массива на 1
        return arr, num  # Вернуть обновленный массив и num

def get_next(start_array, val_m, pos_flag):
    limit = val_m if pos_flag else 1  # Определение лимита на основе pos_flag
    position = len(start_array) - 1  # Определение позиции для обновления
    return array_upd(start_array, position, limit)  # Обновление массива

def build_chain_recursive(chain, array_of_idx, current_index=0):
    if current_index == len(array_of_idx):  # Если текущий индекс равен длине массива array_of_idx
        return chain  # Вернуть текущую цепочку
    else:
        # Добавить новое значение в цепочку на основе array_of_idx
        chain.append(chain[-1] + chain[array_of_idx[current_index] - 1])
        # Рекурсивный вызов с обновленным индексом
        return build_chain_recursive(chain, array_of_idx, current_index + 1)

def line_create(array_of_idx):
    initial_chain = [1]  # Инициализация начальной цепочки
    return build_chain_recursive(initial_chain, array_of_idx)  # Строить цепочку рекурсивно

def arr_dv(n, m_std=None):
    # Вложенная функция для поиска цепочки с заданными параметрами
    def find_chain(array_of_r, array_of_p, q, m):
        num_id = 0
        while num_id != -1:
            cur_line = line_create(array_of_r + array_of_p)  # Создание текущей линии
            max1 = cur_line[q] * (2 ** (m - q))  # Вычисление максимального значения
            min1 = cur_line[q] + (m - q)  # Вычисление минимального значения

            if n < min1 or n > max1:  # Если n не входит в допустимый диапазон
                array_of_r, num_id = get_next(array_of_r, q, False)  # Обновление массива array_of_r
                continue

            num_id = m
            while num_id != -1:
                cur_line = line_create(array_of_r + array_of_p)  # Создание текущей линии
                if n == cur_line[-1]:  # Если n равно последнему элементу текущей линии
                    return cur_line  # Вернуть текущую линию

                array_of_p, num_id = get_next(array_of_p, m - q, True)  # Обновление массива array_of_p

        array_of_r, num_id = get_next(array_of_r, q, False)  # Обновление массива array_of_r
        array_of_p = []  # Очистка массива array_of_p
        for pr in range(1, m - q + 1):  # Заполнение массива array_of_p
            array_of_p.append(pr + q)
    if n == 1:  # Если n равно 1
        return [1]  # Вернуть тривиальную цепочку [1]
    for m in range(int(log2(n)), int(log2(n)) + list(bin(n)).count('1') - 1 + 1):  # Цикл по возможным значениям m
        if m_std:  # Если задано значение m_std
            m = m_std  # Установить m равным m_std
        Q = m // 2  # Вычисление значения Q
        if Q == 0: Q = 1  # Если Q равно 0, установить значение Q равным 1
        array_of_r = []  # Инициализация массива array_of_r
        for r in range(1, Q + 1):  # Заполнение массива array_of_r
            array_of_r.append(r)
        array_of_p = []  # Инициализация массива array_of_p
        for p in range(1, m - Q + 1):  # Заполнение массива array_of_p
            array_of_p.append(Q + p)
        answer = find_chain(array_of_r, array_of_p, Q, m)  # Поиск цепочки с заданными параметрами
        if answer:  # Если найдена цепочка
            return answer  # Вернуть найденную цепочку
    return [] 


### Тестировка и время исполнения алгоритма

In [4]:
import timeit

input_array=[1001,1010,1024,1026,1050]
for i in input_array:
    elapsed_time = timeit.timeit(lambda: arr_dv(i), number=1)
    print("Полученная аддитивная цепочка: ", arr_dv(i))
    print("Время выполнения: {:.6f} секунд".format(elapsed_time))

Полученная аддитивная цепочка:  [1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000, 1001]
Время выполнения: 15.432662 секунд
Полученная аддитивная цепочка:  [1, 2, 4, 8, 16, 32, 64, 128, 256, 320, 336, 672, 1008, 1010]
Время выполнения: 12.116002 секунд
Полученная аддитивная цепочка:  [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]
Время выполнения: 0.000056 секунд
Полученная аддитивная цепочка:  [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1026]
Время выполнения: 0.001737 секунд
Полученная аддитивная цепочка:  [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1040, 1048, 1050]
Время выполнения: 11.645670 секунд


| Number n    | len(chain) | Chain                                                         | time, sec |
|------|---------------|-----------------------------------------------------------------|--------------|
| 1001 | 14            | [1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 400, 800, 1000, 1001]   | 15.297053     |
| 1010 | 14            | [1, 2, 4, 8, 16, 32, 64, 128, 256, 320, 336, 672, 1008, 1010]   | 12.070393     |
| 1024 | 11            | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024]                   | 0.000161     |
| 1026 | 12            | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1026]             | 0.002639     |
| 1050 | 14            | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1040, 1048, 1050] | 11.211253     |

### Вывод
Алгоритм занимает много времени на выполнение, но взамен предоставляет абсолютно минимальную звездную цепочку, обеспечивая точность результата. Такой подход, основанный на полном переборе, приводит к тому, что алгоритм становится менее эффективным при решении задач с большими значениями n, что может затруднить его использование в реальных приложениях или для быстрого получения ответа. Однако, несмотря на низкую производительность, этот алгоритм может быть полезен в некоторых случаях, когда требуется гарантированная точность и минимальная длина звездной цепочки. Возможно, с применением оптимизаций или использованием мощных вычислительных ресурсов, процесс поиска минимальной звездной цепочки с помощью этого алгоритма может быть ускорен, хотя его эффективность по-прежнему будет ограничена при больших n.

### Гипотеза Шольца-Брауэра

Для того чтобы проверить гипотезу о дополнении вектора в алгоритме, рассмотрим случай для числа $2^{12}-1$. Основываясь на предположении, что для всех значений $n \leq 64$ справедливо равенство длин, мы можем в два раза подробнее исследовать этот аспект алгоритма. Такой подход позволяет более глубоко оценить применимость и достоверность гипотезы для данного случая, учитывая особенности алгоритма и ограничения, связанные с допустимыми значениями $n$. Это поможет определить, насколько устойчиво равенство длин выполняется при условии указанных ограничений и возможных изменений в алгоритме или применяемых методах.

In [5]:
int_fl=1
kl = 0
for num in range(1, 12 + 1):
    l_r_eq= len(arr_dv(num)) + num - 1
    a = 2 ** num -1
    b = l_r_eq - 1
    res = arr_dv(a, b)
    if len(res)<=l_r_eq:
        print(f"Гипотеза верна при n = {num}")
        kl+=1
    else:
        print(f"Гипотеза неверна при n = {num}")
        break
        
if kl == 12:
    print("Доказано")
else:
    print("Не доказано")

Гипотеза верна при n = 1
Гипотеза верна при n = 2
Гипотеза верна при n = 3
Гипотеза верна при n = 4
Гипотеза верна при n = 5
Гипотеза верна при n = 6
Гипотеза верна при n = 7
Гипотеза верна при n = 8
Гипотеза верна при n = 9
Гипотеза верна при n = 10
Гипотеза верна при n = 11
Гипотеза верна при n = 12
Доказано


### Вывод
Гипотеза Шольца-Брауэра была подвержена проверке, однако этот процесс потребовал значительного улучшения алгоритма дробления вектора. Это было необходимо, поскольку для значения $n=11$ алгоритм уже не мог обеспечить приемлемое время выполнения. Это указывает на то, что решение данной проблемы может потребовать разработки новых алгоритмических подходов и использования более мощных вычислительных ресурсов.

## Выводы

В рамках исследования аддитивных цепочек и их частного случая - звездных цепочек, были разработаны различные алгоритмы для их построения. Однако, проведенный анализ показал, что построение минимальной цепочки остается непростой задачей, и до сих пор нет эффективного алгоритма, который был бы точен и при этом имел бы разумную временную сложность.

Данная проблема имеет важное практическое значение, поскольку минимальные аддитивные цепочки используются в различных математических задачах, например, в криптографии, и могут значительно ускорить процесс вычислений. Поэтому, поиск новых подходов и методов для решения этой задачи является актуальным направлением исследований в области теории чисел.